authentication/src/test/java/io/camunda/authentication/filters/OAuth2RefreshTokenFilterTest.java

```diff
@@ -0,0 +1,464 @@
+        .isInstanceOf(OAuth2AuthenticationException.class)
+        .hasMessageContaining("Access token expired");
+
+    assertThat(authenticationToken.isAuthenticated()).isFalse();
+    verify(logoutHandler).logout(request, response, authenticationToken);
+  }
+
+  @Test
+  void shouldUseCachedTokenWhenAvailableAndNotExpired() throws ServletException, IOException {
+    // Given
+    final OAuth2AuthorizedClient expiredClient = createExpiredAuthorizedClientWithRefreshToken();
+    final OAuth2AuthorizedClient refreshedClient = createValidAuthorizedClient();
+
+    when(securityContext.getAuthentication()).thenReturn(authenticationToken);
+    when(authorizedClientService.loadAuthorizedClient(CLIENT_REGISTRATION_ID, PRINCIPAL_NAME))
+        .thenReturn(expiredClient);
+    when(authorizedClientManager.authorize(any(OAuth2AuthorizeRequest.class)))
+        .thenReturn(refreshedClient);
+
+    // first request should refresh and cache
+    filter.doFilterInternal(request, response, filterChain);
+
+    // second request should use cache
+    filter.doFilterInternal(request, response, filterChain);
+
+    // Then
+    verify(filterChain, times(2)).doFilter(request, response);
+    verify(authorizedClientManager, times(1)).authorize(any(OAuth2AuthorizeRequest.class));
+  }
+
+  @Test
+  void shouldHandleConcurrentRequestsForSamePrincipal() throws Exception {
+    // Given
+    final OAuth2AuthorizedClient expiredClient = createExpiredAuthorizedClientWithRefreshToken();
+    final OAuth2AuthorizedClient refreshedClient = createValidAuthorizedClient();
+
+    when(securityContext.getAuthentication()).thenReturn(authenticationToken);
+    when(authorizedClientService.loadAuthorizedClient(CLIENT_REGISTRATION_ID, PRINCIPAL_NAME))
+        .thenReturn(expiredClient);
+    when(authorizedClientManager.authorize(any(OAuth2AuthorizeRequest.class)))
+        .thenReturn(refreshedClient);
+
+    // When
+    // simulating concurrent requests
+    final Thread thread1 =
+        new Thread(
+            () -> {
+              try {
+                filter.doFilterInternal(request, response, filterChain);
+              } catch (final Exception e) {
+                throw new RuntimeException(e);
+              }
+            });
+
+    final Thread thread2 =
+        new Thread(
+            () -> {
+              try {
+                filter.doFilterInternal(request, response, filterChain);
+              } catch (final Exception e) {
+                throw new RuntimeException(e);
+              }
+            });
+
+    thread1.start();
+    thread2.start();
+    thread1.join();
+    thread2.join();
+
+    // Then
+    verify(filterChain, times(2)).doFilter(request, response);
+    // should only call authorize once
+    verify(authorizedClientManager, times(1)).authorize(any(OAuth2AuthorizeRequest.class));
+  }
+
+  @Test
+  void shouldReturnNullForNonOAuth2AuthenticationToken() {
```

Comment:
I think this test is a bit more robust and valuable if it tests the behavior of the `doFilterInternal` method and not just the getter. Testing with the getter puts a lot of assumptions about the internal behavior of the class. Of course testing `doFilterInternal` does too, but to a lesser degree as `doFilterInternal` implements the contract of the class.

Reply:
Sure, done